src/main/scala/com/fulcrumgenomics/basecalling/ExtractBasecallingParams.scala

```diff
@@ -0,0 +1,191 @@
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.basecalling
+
+import com.fulcrumgenomics.FgBioDef.{DirPath, FilePath, PathToBam, unreachable}
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.Io
+import com.fulcrumgenomics.util.miseq.{Sample, SampleSheet}
+import dagr.commons.io.PathUtil
+import dagr.commons.util.LazyLogging
+import dagr.sopt.{arg, clp}
+
+import scala.collection.mutable.ListBuffer
+
+
+@clp(group=ClpGroups.Basecalling, description=
+  """Extracts sample and library information from an sample sheet for a given lane.
+    |
+    |The sample sheet should be an Illumina Experiment Manager sample sheet. The tool writes two files to the output
+    |directory: a barcode parameter file and a library parameter file.
+    |
+    |The barcode parameter file is used by Picard's ExtractIlluminaBarcodes and CollectIlluminaBasecallingMetrics to
+    |determine how to match sample barcodes to each read.  The parameter file will be written to the output directory
+    |with name "barcode_params.lane-<lane>.txt".
+    |
+    |The library parameter file is used by Picard's IlluminaBasecallsToSam to demultiplex samples and name each sample's
+    |output BAM file.  The parameter file will be written to the output directory with name
+    |"library_params.lane-<lane>.txt".  The path to each sample's BAM file will be specified in the library parameter
+    |file.  Each BAM file will have path "<output>/<sample-name>.<library-id>.bam" if the library identifier is given or
+    |"<output>/<sample-name>.bam" if library identifer is not specified in the sample sheet.
+  """)
+class ExtractBasecallingParams
+(
+  @arg(flag="i", doc="The sample sheet.") val sampleSheet: FilePath,
+  @arg(flag="o", doc="The output folder to where lane-specific data should be written.") val output: DirPath,
```

Comment:
I think probably there should be a second parameter, something like `bamOutput: Option[DirPath]` that defaults to `output` at runtime, but allows users to specify a different directory to output BAMs to.